### PR TITLE
Fix crash in overload resolution with tuple literals

### PR DESCRIPTION
**Pull request by jaredpar**
The overload resolution code had a bad dependency on `BoundExpression.Display` being a `TypeSymbol`.  That member appears to be added specifically for the opposite case.  Essentially `BoundExpression` values which can't possibly produce a valid `TypeSymbol`: null literals, method groups, etc ... Over time it's also become a place for `BoundExpression` nodes to customize their display.

The bad dependency has gone unnoticed up until now because of previous error handling in overload resolution.  All of the `BoundExpression` which can't produce a type were special cased in one formm or another earlier on.

Our work in tuple litercals exposed this bug though.  It added an override for `BoundExpression.Display` which is a `string` (completely valid) and no previous special casing existed for this type and it ended up hitting the code with the bad dependency.

Fix is to properly handle the case wherce the value is a not a `TypeSymbol`.

This addresses DevDiv 258853 (Watson bug)

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @MattGertz for approval

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler for code review

**Comment by VSadov** [comment]
I wonder if the crash was reachable with lambdas as well. - they have no natural type too.

**Comment by VSadov** [comment]
LGTM

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb [hunk 1]
```diff
@@ -7161,6 +7161,59 @@ additionalReferences:=s_valueTupleRefs)
...
+    Sub Main()
+        M((1, 2))
+    End Sub
+End Module
+
+    </file>
+</compilation>, additionalRefs:=s_valueTupleRefs)
+
+
+            compilation.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: >VerifyDiagnostics [](start = 24, length = 17)

Please use AssertTheseDiagnostics helper instead.

**Comment by jcouv** [comment]
LGTM
Do we have a similar issue on the VB side?

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb [hunk 2]
```diff
@@ -7161,6 +7161,59 @@ additionalReferences:=s_valueTupleRefs)
...
+    Sub Main()
+        M((1, Nothing))
+    End Sub
+End Module
+
+    </file>
+</compilation>, additionalRefs:=s_valueTupleRefs)
+
+
+            compilation.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: >VerifyDiagnostics [](start = 24, length = 17)

Here as well.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by jaredpar** [comment]
@jcouv there was no bug I could find.  I added the unit tests as a precaution.

**Comment by jaredpar** [comment]
CC @Pilchie for ask mode approval

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb [hunk 3]
```diff
@@ -7161,6 +7161,59 @@ additionalReferences:=s_valueTupleRefs)
...
+<compilation>
+    <file name="a.vb">
+Module C
+    Sub M(x As Integer)
+    End Sub
+    Sub M(x As String)
+    End Sub
+
+    Sub Main()
+        M((1, Nothing))
```
Comment (VSadov) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: Try  a case  with ``` M((1, AddressOf Main)) ```

AddressOf truly does not have type in VB, while Nothing, kindof, does - it has a default type of Object.

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: Will do.

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb (hunk 3 above)
Comment (jaredpar) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: Will add that case.

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb [hunk 4]
```diff
@@ -7161,6 +7161,98 @@ additionalReferences:=s_valueTupleRefs)
...
+End Module
+
+    </file>
+</compilation>, additionalRefs:=s_valueTupleRefs)
+
+
+            compilation.AssertTheseDiagnostics(
+<errors>
+    BC30518: Overload resolution failed because no accessible 'M' can be called with these arguments:
+    'Public Sub M(x As Integer)': Expression does not produce a value.
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTuples.vb]: >Public Sub M(x As Integer)': Expression does not produce a value. [](start = 5, length = 65)

This shouldn't block this PR, but it feels like we should be getting a friendlier message here. Something about AddressOf.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
:shipit:

**Comment by Pilchie** [comment]
Approved

**Comment by jcouv** [comment]
For reference, I resolved Watson 269814 as a duplicate of this issue (nullref with same stacktrace, my investigation led me back to this fix).

**Pull request merged**